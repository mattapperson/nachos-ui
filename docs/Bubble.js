import React from "react";
import { View } from "react-native";

import Bubble from "../src/Bubble";
import H4 from "../src/typography/H4";

// WEBPACK (BubbleExample)
const BubbleExample = () => {
  return (
    <View>
      <H4>Example:</H4>
      <View style={{ marginVertical: 15, flexDirection: "row" }}>
        <Bubble
          arrowPosition="left"
          bubbleStyle={{
            backgroundColor: "#ececee"
          }}
          textStyle={{
            color: "#8a8aa2"
          }}
        >
          What’s cooking?
        </Bubble>
        <Bubble
          arrowPosition="right"
          style={{
            marginTop: 60,
            marginLeft: 10
          }}
        >
          Hey, What’s Up?
        </Bubble>
      </View>
    </View>
  );
};
// WEBPACK

const description = `
| Prop | Type | Default | Description |
|------|------|---------|-------------|
| \`arrowPosition\` | \`enum\` | \`left\` | Determines the position of the arrow. One of: \`'top', 'left', 'bottom', 'right'\` |
| \`color\` | \`string\` | \`'#2f8cff'\` | Background color of the Bubble compoment |
| \`style\` | \`object, array\` | \`void\` | Style of the Bubble compoment |
| \`theme\` | \`object\` | \`defaultTheme\` | An object describing a style of the component. See more in Theme section.  |
`;

BubbleExample.styleguide = {
  ...BubbleExample.styleguide,
  title: "Bubble",
  description
};

export default BubbleExample;
