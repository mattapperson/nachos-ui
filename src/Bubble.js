import React from 'react'
import PropTypes from 'prop-types'
import { View, Text } from 'react-native'

const defaultTheme = {
  BUBBLE_BACKGROUND: '#2f8cff',
  BUBBLE_MIN_HEIGHT: 70,
  BUBBLE_BORDER_RADIUS: 10,
  BUBBLE_PADDING_VERTICAL: 15,
  BUBBLE_PADDING_HORIZONTAL: 25,
  BUBBLE_FONT_SIZE: 14,
  BUBBLE_FONT_COLOR: '#fff',
}

const Bubble = (props) => {
  const {
    arrowPosition,
    style,
    children,
  } = props

  const theme = props.theme || themeManager.getStyle('Bubble')
  const baseStyle = defaultStyle(theme)
  const color = props.color || theme.BUBBLE_BACKGROUND

  return (
    <View style={style}>
      <View style={[baseStyle.base, { backgroundColor: color }]}>
        <Text style={baseStyle.text}>
          {children}
        </Text>
        <View
          style={[
            baseStyle.arrowContainer,
            baseStyle.arrowPosition[arrowPosition],
          ]}
        >
          <View style={baseStyle.arrow(color)[arrowPosition]} />
        </View>
      </View>
    </View>
  )
}

Bubble.defaultStyle = (theme = defaultTheme) => {
  return {
    base: {
      position: 'relative',
      minHeight: theme.BUBBLE_MIN_HEIGHT,
      borderRadius: theme.BUBBLE_BORDER_RADIUS,
      flexDirection: 'column',
      alignItems: 'center',
      justifyContent: 'center',
      paddingVertical: theme.BUBBLE_PADDING_VERTICAL,
      paddingHorizontal: theme.BUBBLE_PADDING_HORIZONTAL,
    },
    text: {
      color: theme.BUBBLE_FONT_COLOR,
      fontSize: theme.BUBBLE_FONT_SIZE,
    },
    arrowContainer: {
      position: 'absolute',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
    },
    arrowPosition: {
      top: {
        justifyContent: 'flex-start',
        alignItems: 'center',
      },
      right: {
        justifyContent: 'center',
        alignItems: 'flex-end',
      },
      bottom: {
        justifyContent: 'flex-end',
        alignItems: 'center',
      },
      left: {
        justifyContent: 'center',
        alignItems: 'flex-start',
      },
    },
    arrow: (color) => {
      return {
        top: {
          marginTop: -7,
          borderRightWidth: 6,
          borderBottomWidth: 7,
          borderLeftWidth: 6,
          borderRightColor: 'transparent',
          borderBottomColor: color,
          borderLeftColor: 'transparent',
        },
        right: {
          marginRight: -7,
          borderTopWidth: 6,
          borderBottomWidth: 6,
          borderLeftWidth: 7,
          borderTopColor: 'transparent',
          borderBottomColor: 'transparent',
          borderLeftColor: color,
        },
        bottom: {
          marginBottom: -7,
          borderRightWidth: 6,
          borderTopWidth: 7,
          borderLeftWidth: 6,
          borderRightColor: 'transparent',
          borderTopColor: color,
          borderLeftColor: 'transparent',
        },
        left: {
          marginLeft: -7,
          borderTopWidth: 6,
          borderRightWidth: 7,
          borderBottomWidth: 6,
          borderTopColor: 'transparent',
          borderRightColor: color,
          borderBottomColor: 'transparent',
        },
      }
    },
  }
}

Bubble.propTypes = {
  arrowPosition: PropTypes.oneOf(['top', 'left', 'bottom', 'right']),
  color: PropTypes.string,
  style: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  children: PropTypes.node.isRequired,
  theme: PropTypes.object,
}

Bubble.defaultProps = {
  arrowPosition: 'left',
}

export default Bubble
