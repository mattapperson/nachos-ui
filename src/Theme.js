import React, { Component } from "react";
import PropTypes from "prop-types";
import { StyleSheet } from "react-native";
import defaultBranding from "./branding_defaults";

export const ThemeContext = React.createContext({
  theme: {}
});

export class Provider extends React.Component {
  state = {};
  static getDerivedStateFromProps(nextProps, prevState) {
    if (JSON.stringify(nextProps.theme) !== JSON.stringify(prevState)) {
      //return this.createTheme(nextProps.theme);
    }
    return null;
  }

  themeConfigs = {};
  branding = defaultBranding;

  updateThemeConfig = (name, themeConfig) => {
    themeConfig = JSON.parse(
      JSON.stringify(themeConfig).replace(/@([\w_-]+)/gm, (match, key) => {
        if (!this.branding[key]) console.log(key);

        return this.branding[key];
      })
    );

    // if (name === "H1") {
    //   console.log(themeConfig);
    // }
    return (this.themeConfigs[name] = {
      computedStyle: themeConfig.style
        ? StyleSheet.create(themeConfig.style)
        : {},
      props: themeConfig.settings
    });
  };

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.themeConfigs,
          updateThemeConfig: this.updateThemeConfig
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export function withTheme(componentName, ThemedComponent) {
  return class extends React.Component {
    constructor() {
      super();
      try {
        ThemeContext._currentValue.updateThemeConfig(
          componentName,
          ThemedComponent.themeConfig
        );
      } catch (e) {
        console.warn(`failed to update theme for ${componentName}`, e);
      }
    }

    isObject(item) {
      return item && typeof item === "object" && !Array.isArray(item);
    }

    mergeDeep(target, ...sources) {
      if (!sources.length) return target;
      const source = sources.shift();

      if (this.isObject(target) && this.isObject(source)) {
        for (const key in source) {
          if (this.isObject(source[key])) {
            if (!target[key]) Object.assign(target, { [key]: {} });
            this.mergeDeep(target[key], source[key]);
          } else {
            Object.assign(target, { [key]: source[key] });
          }
        }
      }

      return this.mergeDeep(target, ...sources);
    }
    render() {
      return (
        <ThemeContext.Consumer>
          {({ theme }) => {
            const computedTheme =
              (theme[componentName] || {}).computedStyle || {};

            const fullTheme = this.props.theme
              ? this.mergeDeep({}, computedTheme, this.props.theme)
              : computedTheme;

            return (
              <ThemedComponent
                {...(theme[componentName] || {}).props || {}}
                {...this.props}
                theme={fullTheme}
              />
            );
          }}
        </ThemeContext.Consumer>
      );
    }
  };
}
