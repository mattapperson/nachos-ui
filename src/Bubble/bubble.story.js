import { storiesOf } from "../storybook";
import { withInfo } from "../storybook";
import React from "react";
import Bubble from "./index";
import { View } from "react-native";

const Example = () => {
  return (
    <View style={{ marginVertical: 15, flexDirection: "row" }}>
      <Bubble
        arrowPosition="right"
        bubbleStyle={{
          backgroundColor: "#ececee"
        }}
        textStyle={{
          color: "#8a8aa2"
        }}
      >
        Whatâ€™s cooking?
      </Bubble>
    </View>
  );
};

storiesOf("Bubble", module).add("default", withInfo()(Example));

export default Example;
