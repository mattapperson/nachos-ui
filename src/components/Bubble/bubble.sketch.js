import { storiesOf } from "../../storybook";
import { withInfo } from "../../storybook";
import React from "react";
import Bubble from "./index";
import { View } from "react-native";

export const components = [
  {
    Component: () => (
      <Bubble
        bubbleStyle={{
          backgroundColor: "#ececee"
        }}
        textStyle={{
          color: "#8a8aa2"
        }}
      >
        What’s cooking?
      </Bubble>
    ),
    name: "Bubble/Left"
  },
  {
    Component: () => <Bubble arrowPosition="right">What’s cooking?</Bubble>,
    name: "Bubble/Right"
  }
];
