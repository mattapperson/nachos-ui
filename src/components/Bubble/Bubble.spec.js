import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import Bubble from "../Bubble";
import { Provider } from "../Theme";

it("should render correctly", () => {
  const component = renderer.create(
    <Provider>
      <Bubble>Hey, Whatâ€™s Up?</Bubble>
    </Provider>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
