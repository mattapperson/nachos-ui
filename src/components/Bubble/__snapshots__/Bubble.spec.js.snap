// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
>
  <div
    className="rn-alignItems-1awozwy rn-backgroundColor-80k8qj rn-borderTopLeftRadius-j9b53g rn-borderTopRightRadius-1i97xy8 rn-borderBottomRightRadius-tabonr rn-borderBottomLeftRadius-shm4j rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-justifyContent-1777fci rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-1a1dyw rn-minWidth-bcqeeo rn-paddingTop-1qxgc49 rn-paddingRight-mzjzbw rn-paddingBottom-19yat4t rn-paddingLeft-b2dl2 rn-position-bnwqim rn-zIndex-1lgpqti"
  >
    <div
      className="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-1471scf rn-fontFamily-baq3cm rn-fontSize-1b43r93 rn-fontStyle-13uqrnb rn-fontVariant-ebii48 rn-fontWeight-aruset rn-height-1pexk7n rn-lineHeight-rjixqe rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-m611by rn-paddingRight-9aemit rn-paddingBottom-1mi0q7o rn-paddingLeft-gy4na3 rn-textAlign-fdjqy7 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
      dir="auto"
      style={
        Object {
          "color": "rgba(255,255,255,1.00)",
        }
      }
    >
      Hey, Whatâ€™s Up?
    </div>
    <div
      className="rn-alignItems-1habvwh rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-bottom-1p0dtai rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-justifyContent-1777fci rn-left-1d2f490 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-u8s1d rn-right-zchlnj rn-top-ipm5af rn-zIndex-1lgpqti"
    >
      <div
        className="rn-alignItems-1oszu61 rn-borderBottomColor-oucylx rn-borderBottomWidth-1ppbya7 rn-borderRightWidth-139lwf2 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopColor-1c1gj4h rn-borderTopWidth-12tzs1u rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-1x5g5r8 rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-zIndex-1lgpqti"
        style={
          Object {
            "borderRightColor": "rgba(117,64,238,1.00)",
          }
        }
      />
    </div>
  </div>
</div>
`;
